include/linux/hax_interface_linux.h

```diff
@@ -34,42 +34,17 @@
 
 #include <asm/ioctl.h>
 
-/* The mac specific interface to qemu because of mac's
- * special handling like hax tunnel allocation etc */
-/* HAX model level ioctl */
-#define HAX_IOCTL_VERSION _IOWR(0, 0x20, struct hax_module_version)
-#define HAX_IOCTL_CREATE_VM _IOWR(0, 0x21, uint32_t)
-#define HAX_IOCTL_DESTROY_VM _IOW(0, 0x22, uint32_t)
-#define HAX_IOCTL_CAPABILITY _IOR(0, 0x23, struct hax_capabilityinfo)
-#define HAX_IOCTL_SET_MEMLIMIT _IOWR(0, 0x24, struct hax_set_memlimit)
-
-// Only for backward compatibility with old Qemu.
-#define HAX_VM_IOCTL_VCPU_CREATE_ORIG _IOR(0, 0x80, int)
-
-#define HAX_VM_IOCTL_VCPU_CREATE _IOWR(0, 0x80, uint32_t)
-#define HAX_VM_IOCTL_ALLOC_RAM _IOWR(0, 0x81, struct hax_alloc_ram_info)
-#define HAX_VM_IOCTL_SET_RAM _IOWR(0, 0x82, struct hax_set_ram_info)
-#define HAX_VM_IOCTL_VCPU_DESTROY _IOR(0, 0x83, uint32_t)
-#define HAX_VM_IOCTL_ADD_RAMBLOCK _IOW(0, 0x85, struct hax_ramblock_info)
-#define HAX_VM_IOCTL_SET_RAM2 _IOWR(0, 0x86, struct hax_set_ram_info2)
-#define HAX_VM_IOCTL_PROTECT_RAM _IOWR(0, 0x87, struct hax_protect_ram_info)
-
-#define HAX_VCPU_IOCTL_RUN _IO(0, 0xc0)
-#define HAX_VCPU_IOCTL_SET_MSRS _IOWR(0, 0xc1, struct hax_msr_data)
-#define HAX_VCPU_IOCTL_GET_MSRS _IOWR(0, 0xc2, struct hax_msr_data)
-
-#define HAX_VCPU_IOCTL_SET_FPU _IOW(0, 0xc3, struct fx_layout)
-#define HAX_VCPU_IOCTL_GET_FPU _IOR(0, 0xc4, struct fx_layout)
-
-#define HAX_VCPU_IOCTL_SETUP_TUNNEL _IOWR(0, 0xc5, struct hax_tunnel_info)
-#define HAX_VCPU_IOCTL_INTERRUPT _IOWR(0, 0xc6, uint32_t)
-#define HAX_VCPU_SET_REGS _IOWR(0, 0xc7, struct vcpu_state_t)
-#define HAX_VCPU_GET_REGS _IOWR(0, 0xc8, struct vcpu_state_t)
-
-/* API 2.0 */
-#define HAX_VM_IOCTL_NOTIFY_QEMU_VERSION _IOW(0, 0x84, struct hax_qemu_version)
-
-#define HAX_IOCTL_VCPU_DEBUG _IOW(0, 0xc9, struct hax_debug_t)
+#define HAX_IOCTL_HAX_IO(access, code, type) \
+    _IO(0, code)
```

Comment:
It seems this ioctl *code* (aka. *type*) is used only as sanity-check to ensure that the driver listening to ioctl's coming from a file descriptor is the correct one. In the case of HAXM, we are the ones creating the devices, and we assign them a rather unique name so it's unlikely that another driver accidentally overrides our file operations, or that userland applications submit wrong ioctl's to our device.

Still, I agree sanity checks are convenient, so I'll add a non-trivial type (e.g. what about `'H'`?), and the corresponding checks in the ioctl handler function.
```c
if (_IOC_TYPE(ioctl) != HAX_DEVICE_TYPE)
    return -EINVAL;
```

Reply:
Actually, during the grace period it would look like:
```c
if ((_IOC_TYPE(ioctl) != HAX_DEVICE_TYPE) && (_IOC_TYPE(ioctl) != 0 /* Legacy API */))
    return -EINVAL;
```

Reply:
Regarding the choice of `H`, I just want to point out that `H` is already heavily used by other drivers according to the table in `ioctl-number.txt`. I do see the point that those drivers will never open the HAXM device file, but the document also talks about other (minor) benefits of having unique IOCTL numbers. Anyway, you are the Linux driver maintainer, so you have the final say :-) For Darwin, we may not have guidelines on how to pick an IOCTL _group_, so `H` should be fine.